qualtran/bloqs/rotations/zpow_via_phase_gradient.py

```diff
@@ -0,0 +1,121 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import cast
+
+import sympy
+from attrs import frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    DecomposeTypeError,
+    QBit,
+    QFxp,
+    Signature,
+    Soquet,
+    SoquetT,
+)
+from qualtran.bloqs.arithmetic import XorK
+from qualtran.bloqs.rotations.phase_gradient import AddIntoPhaseGrad
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.resource_counting.generalizers import ignore_alloc_free
+from qualtran.symbolics import ceil, is_symbolic, log2, pi, SymbolicFloat, SymbolicInt
+
+
+@frozen
+class ZPowConstViaPhaseGradient(Bloq):
+    r"""Apply an $ZPow(t)$ using a phase gradient state."""
```

Comment:
Add a brief explanation of the technique, along with an explanation of costs in the docstrings - How many qubits? How many Toffolis? 

Also add a reference

Reply:
added reference

Reply:
Also added qubit and toffoli cost